validator_client/src/doppelganger_service.rs

```diff
@@ -67,95 +321,1026 @@ impl<T: 'static + SlotClock, E: EthSpec> DoppelgangerService<T, E> {
+                    validator client",
+                "doppelganger_indices" => ?violators
+            )
+        }
+
+        // The concept of "epoch satisfaction" is that for some epoch `e` we are *satisfied* that
+        // we've waited long enough such that we don't expect to see any more consensus messages
+        // for that epoch.
+        //
+        // As it stands now, we consider epoch `e` to be satisfied once we're in the last slot of
+        // epoch `e + 1`.
+        //
+        // The reasoning for this choice of satisfaction slot is that by this point we've
+        // *probably* seen all the blocks that are permitted to contain attestations from epoch `e`.
+        let previous_epoch_satisfaction_slot = previous_epoch
+            .saturating_add(1_u64)
+            .end_slot(E::slots_per_epoch());
+        let previous_epoch_is_satisfied = request_slot >= previous_epoch_satisfaction_slot;
+
+        // Iterate through all the previous epoch responses, updating `self.doppelganger_states`.
+        //
+        // Do not bother iterating through the current epoch responses since they've already been
+        // checked for violators and they don't result in updating the state.
+        for response in &previous_epoch_responses {
+            // Sanity check response from the server.
+            //
+            // Abort the entire routine if the server starts returning junk.
+            if response.epoch != previous_epoch {
+                return Err(format!(
+                    "beacon node returned epoch {}, expecting {}",
+                    response.epoch, previous_epoch
+                ));
+            }
+
+            // Resolve the index from the server response back to a public key.
+            let pubkey = indices_map
+                .get(&response.index)
+                // Abort the routine if inconsistency is detected.
+                .ok_or_else(|| {
+                    format!(
+                        "inconsistent indices map for validator index {}",
+                        response.index
+                    )
+                })?;
+
+            // Hold the lock on `self` for the rest of this function.
+            //
+            // !! IMPORTANT !!
+            //
+            // There is a write-lock being held, avoid interacting with locks until it is dropped.
+            let mut doppelganger_states = self.doppelganger_states.write();
+            let doppelganger_state = doppelganger_states
+                .get_mut(&pubkey)
+                // Abort the routine if inconsistency is detected.
+                .ok_or_else(|| format!("inconsistent states for validator pubkey {}", pubkey))?;
+
+            // If a single doppelganger is detected, enable doppelganger checks on all
+            // validators forever (technically only 2**64 epochs).
+            //
+            // This has the effect of stopping validator activity even if the validator client
+            // fails to shut down.
```

Comment:
Rather than setting this to `u64::MAX` could we just increment `next_check_epoch` and reset `remaining_epochs` to `DEFAULT_REMAINING_EPOCHS`?  I can't imagine hitting this weird scenario where doppelganger fails, doesn't shutdown and the VC runs for 2**64 epochs... but it makes more sense to me that you'd just keep pushing back doppelganger detection if for some reason we detected a doppelganger but didn't shut down.

Reply:
> I can't imagine hitting this weird scenario where doppelganger fails, doesn't shutdown and the VC runs for 2**64 epochs... but it makes more sense to me that you'd just keep pushing back doppelganger detection if for some reason we detected a doppelganger but didn't shut down.

By my calculations, 2**64 epochs is 224 trillion years so it's pretty safe. We can't really do any better than this either  since `next_check_epoch` is ultimately a `u64`.

If we went with the push-back method, it seems to me like that has a possibility of recovering. E.g., someone turns off the remote doppelganger and then we'll start working again. Is that what you were suggesting?

Reply:
I was thinking if we can't shut down maybe we should try to recover... but I guess it's still too much of a risk that the doppelganger just missed some attestations. 

> By my calculations, 2**64 epochs is 224 trillion years so it's pretty safe 

About how long I plan to be staking for ^ :rofl:

Reply:
> I was thinking if we can't shut down maybe we should try to recover... but I guess it's still too much of a risk that the doppelganger just missed some attestations.

Yeah, I'm tempted to leave recovery alone. I'm not sure how to specify it in a way that is safe. I think it's nice to maintain the promise that "if a doppelganger is detected the VC will stop".